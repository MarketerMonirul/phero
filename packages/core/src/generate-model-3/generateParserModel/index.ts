import ts from "typescript"
import { ParseError } from "../../domain/errors"
import { ParserModel } from "../ParserModel"
import generateFromTypeNode from "./generateFromTypeNode"

export interface ParserModelMap {
  root: ParserModel
  deps: Record<string, ParserModel>
}

export interface InternalParserModelMap {
  root: ParserModel
  deps: DependencyMap
}

export type DependencyMap = Map<string, { name: string; model: ParserModel }>
export type TypeParamMap = Map<string, { name: string; model: ParserModel }>

export function generateParserModel(
  func: ts.FunctionDeclaration,
  prog: ts.Program,
): ParserModelMap {
  const typeChecker = prog.getTypeChecker()

  const funcType = func.type

  if (!funcType) {
    throw new ParseError("Function must have type", func)
  }

  const funcTypeType = typeChecker.getTypeAtLocation(
    funcType,
  ) as ts.TypeReference

  const { root, deps } = generateFromTypeNode(
    funcType,
    funcTypeType,
    funcType,
    typeChecker,
    new Map(),
    new Map(),
  )

  return {
    root,
    deps: [...deps.values()].reduce<Record<string, ParserModel>>(
      (result, dep) => ({ ...result, [dep.name]: dep.model }),
      {},
    ),
  }
}
