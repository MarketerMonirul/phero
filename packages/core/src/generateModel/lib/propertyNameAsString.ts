import ts from "typescript"
import { PheroParseError } from "../../domain/errors"

export default function propertyNameAsString(
  propertyName: ts.PropertyName,
): string {
  if (ts.isIdentifier(propertyName)) {
    return propertyName.text
  }
  if (ts.isStringLiteral(propertyName)) {
    return propertyName.text
  }
  if (ts.isNumericLiteral(propertyName)) {
    return propertyName.text
  }
  if (ts.isComputedPropertyName(propertyName)) {
    throw new PheroParseError(
      "Member name must not be computed property",
      propertyName,
    )
  }

  if (ts.isPrivateIdentifier(propertyName)) {
    throw new PheroParseError(
      "Member name must not be private identifier",
      propertyName,
    )
  }

  throw new PheroParseError(`Unexpected value for member name`, propertyName)
}
