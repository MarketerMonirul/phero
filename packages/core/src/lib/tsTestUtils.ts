import ts, {
  FunctionLikeDeclarationBase,
  VariableDeclaration,
} from "typescript"
import { TSFiles, VirtualCompilerHost } from "./VirtualCompilerHost"
import { PheroApp } from "../domain/PheroApp"
import { KindToNodeMappings } from "./tsUtils"
import {
  DependencyMap,
  FunctionParserModel,
  generateParserModel,
  ParserModelMap,
} from "../generate-model-3/generateParserModel"
import { ObjectParserModel, ParserModel } from "../generate-model-3/ParserModel"

export function printPheroApp(app: PheroApp): string {
  return JSON.stringify(
    {
      services: app.services.map((service) => ({
        name: service.name,
        funcs: service.funcs.map((func) => ({
          name: func.name,
          func: printFunctionDeclaration(func.ref),
        })),
      })),
    },
    null,
    4,
  )

  function printFunctionDeclaration(
    func: FunctionLikeDeclarationBase | VariableDeclaration,
  ): string {
    const funcName = func.name?.getText()

    if (!funcName) {
      throw new Error("Func must have name")
    }

    return func.kind === ts.SyntaxKind.FunctionDeclaration
      ? `[FunctionDeclaration(${funcName})]`
      : func.kind === ts.SyntaxKind.VariableDeclaration
      ? `[VariableDeclaration(${funcName})]`
      : func.kind === ts.SyntaxKind.FunctionExpression
      ? `[FunctionExpression(${funcName})]`
      : func.kind === ts.SyntaxKind.ArrowFunction
      ? `[ArrowFunction(${funcName})]`
      : "[UNKNOWN]"
  }
}

export function createTestProgram(input: TSFiles | string): ts.Program {
  const vHost = new VirtualCompilerHost()

  if (typeof input === "string") {
    vHost.addFile(`phero.ts`, input)
  } else {
    for (const [fileName, content] of Object.entries(input)) {
      vHost.addFile(`${fileName}.ts`, content)
    }
  }

  const program = vHost.createProgram("phero.ts")
  return program
}

export function compileStatement<SK extends ts.SyntaxKind>(
  code: string,
  syntaxKind: SK,
): { statement: KindToNodeMappings[SK]; prog: ts.Program } {
  const prog = createTestProgram(code)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (statements?.length !== 1) {
    throw new Error("Should provide exactly 1 statement")
  }
  const statement = statements[0]
  if (statement.kind !== syntaxKind) {
    throw new Error(
      `SytaxKind of statement is ${statement.kind}, but ${syntaxKind} was expected`,
    )
  }

  return {
    prog,
    statement: statement as KindToNodeMappings[SK],
  }
}

export function compileStatements(code: string): {
  statements: ts.Statement[]
  prog: ts.Program
} {
  const prog = createTestProgram(code)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (!statements || statements.length < 1) {
    throw new Error("Should provide at least 1 statement")
  }

  return {
    prog,
    statements: statements.map((s) => s),
  }
}

export function compileProgram(input: TSFiles): {
  statements: ts.Statement[]
  prog: ts.Program
} {
  const prog = createTestProgram(input)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (!statements || statements.length < 1) {
    throw new Error("Should provide at least 1 statement")
  }

  return {
    prog,
    statements: statements.map((s) => s),
  }
}

const printer = ts.createPrinter({
  newLine: ts.NewLineKind.LineFeed,
  noEmitHelpers: true,
  removeComments: true,
  omitTrailingSemicolon: false,
})

export function printCode(code: ts.Node | ts.Node[]): string {
  const sf = ts.createSourceFile(
    "phero-manifest.d.ts",
    "",
    ts.ScriptTarget.ESNext,
    undefined,
    ts.ScriptKind.TS,
  )

  return Array.isArray(code)
    ? printer.printList(
        ts.ListFormat.SourceFileStatements,
        ts.factory.createNodeArray(code),
        sf,
      )
    : printer.printNode(ts.EmitHint.Unspecified, code, sf)
}

export function generateParserModelForReturnType(tsContent: string): {
  root: ParserModel
  deps: Record<string, ParserModel>
} {
  const { statements, prog } = compileStatements(tsContent)

  const func = statements.find((st): st is ts.FunctionDeclaration =>
    ts.isFunctionDeclaration(st),
  )

  if (!func) {
    throw new Error("Ts content doesn't contain any function")
  }

  const funcModel = generateParserModel(func, prog.getTypeChecker(), new Map())
  return {
    root: funcModel.returnType,

    deps: [...funcModel.deps.entries()].reduce<Record<string, ParserModel>>(
      (result, [name, model]) => ({ ...result, [name]: model }),
      {},
    ),
  }
}

export function generateParserModelForFunction(tsContent: string): {
  returnType: ParserModel
  parameters: ObjectParserModel
  deps: Record<string, ParserModel>
} {
  const { statements, prog } = compileStatements(tsContent)

  const func = statements.find((st): st is ts.FunctionDeclaration =>
    ts.isFunctionDeclaration(st),
  )

  if (!func) {
    throw new Error("Ts content doesn't contain any function")
  }

  const funcModel = generateParserModel(func, prog.getTypeChecker(), new Map())

  return {
    returnType: funcModel.returnType,
    parameters: funcModel.parameters,
    deps: [...funcModel.deps.entries()].reduce<Record<string, ParserModel>>(
      (result, [name, model]) => ({ ...result, [name]: model }),
      {},
    ),
  }
}
