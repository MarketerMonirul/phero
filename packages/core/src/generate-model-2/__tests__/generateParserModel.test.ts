import fs from "fs"
import path from "path"
import ts from "typescript"
import { compileStatements } from "../../lib/tsTestUtils"
import generateParserModel from "../generateParserModel"

describe("generateParserModel", () => {
  getExamples()
    .filter(([name]) =>
      [
        // "_date.ts",
        "array.ts",
        // "enum.ts",
        // "generics.ts",
        // "indexType.ts",
        // "indexedAccess.ts",
        "interface.ts",
        // "intersection.ts",
        "literal.ts",
        // "mappedType.ts",
        "native.ts",
        // "objectLiteral.ts",
        // "recursive_data_type.ts",
        // "tuple.ts",
        "typeAlias.ts",
        "union.ts",
      ].includes(name),
    )
    .forEach(([name, tsContent]) => {
      return describe(name, () => {
        const [funcs, prog] = getFunctionDeclarations(tsContent)
        funcs.forEach((func) => {
          const testCallback = (): void => {
            expect(generateParserModel(func, prog)).toMatchSnapshot(
              func.name?.text,
            )
          }
          const funcName = `${func.name?.text ?? "-"}`
          if (funcName.startsWith("_")) {
            test.only(funcName.substring(1), testCallback)
          } else {
            test(funcName, testCallback)
          }
        })
      })
    })
})

function getExamples(): Array<[name: string, example: string]> {
  const examples = fs.readdirSync(path.join(__dirname, "./examples"))
  return examples.map((fileName) => [
    fileName,
    fs.readFileSync(path.join(__dirname, "./examples", fileName), {
      encoding: "utf-8",
    }),
  ])
}

function getFunctionDeclarations(
  tsContent: string,
): [ts.FunctionDeclaration[], ts.Program] {
  const { statements, prog } = compileStatements(tsContent)
  return [
    statements.filter((st): st is ts.FunctionDeclaration =>
      ts.isFunctionDeclaration(st),
    ),
    prog,
  ]
}
