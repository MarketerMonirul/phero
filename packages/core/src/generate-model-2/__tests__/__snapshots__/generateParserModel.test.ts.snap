// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateParserModel array.ts getBooleanArray1: getBooleanArray1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "boolean",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getBooleanArray2: getBooleanArray2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "boolean",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getBooleanDoubleArray1: getBooleanDoubleArray1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "element": Object {
          "parser": Object {
            "type": "boolean",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getBooleanDoubleArray2: getBooleanDoubleArray2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "element": Object {
          "parser": Object {
            "type": "boolean",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getNumberArray1: getNumberArray1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "number",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getNumberArray2: getNumberArray2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "number",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getNumberDoubleArray1: getNumberDoubleArray1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "element": Object {
          "parser": Object {
            "type": "number",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getNumberDoubleArray2: getNumberDoubleArray2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "element": Object {
          "parser": Object {
            "type": "number",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getStringArray1: getStringArray1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "string",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getStringArray2: getStringArray2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "string",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getStringDoubleArray1: getStringDoubleArray1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "element": Object {
          "parser": Object {
            "type": "string",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel array.ts getStringDoubleArray2: getStringDoubleArray2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "element": Object {
          "parser": Object {
            "type": "string",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel interface.ts getInterface1: getInterface1 1`] = `
Object {
  "deps": Object {
    "Message1": Object {
      "members": Array [
        Object {
          "name": "s",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "n",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "b",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "os",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "on",
          "optional": true,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "ob",
          "optional": true,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "nested",
          "optional": false,
          "parser": Object {
            "type": "reference",
            "typeName": "NestedMessage",
          },
          "type": "member",
        },
        Object {
          "name": "oNested",
          "optional": true,
          "parser": Object {
            "type": "reference",
            "typeName": "NestedMessage",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
    "NestedMessage": Object {
      "members": Array [
        Object {
          "name": "x",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "y",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "z",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "ox",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "oy",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "oz",
          "optional": true,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message1",
  },
}
`;

exports[`generateParserModel interface.ts getInterface2: getInterface2 1`] = `
Object {
  "deps": Object {
    "Message1": Object {
      "members": Array [
        Object {
          "name": "s",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "n",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "b",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "os",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "on",
          "optional": true,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "ob",
          "optional": true,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "nested",
          "optional": false,
          "parser": Object {
            "type": "reference",
            "typeName": "NestedMessage",
          },
          "type": "member",
        },
        Object {
          "name": "oNested",
          "optional": true,
          "parser": Object {
            "type": "reference",
            "typeName": "NestedMessage",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
    "NestedMessage": Object {
      "members": Array [
        Object {
          "name": "x",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "y",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "z",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "ox",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "oy",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "oz",
          "optional": true,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "reference",
        "typeName": "Message1",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel interface.ts getInterface3: getInterface3 1`] = `
Object {
  "deps": Object {
    "Message3": Object {
      "members": Array [
        Object {
          "name": "strArr",
          "optional": false,
          "parser": Object {
            "element": Object {
              "parser": Object {
                "type": "string",
              },
              "type": "arrayElement",
            },
            "type": "array",
          },
          "type": "member",
        },
        Object {
          "name": "msgArr",
          "optional": false,
          "parser": Object {
            "element": Object {
              "parser": Object {
                "type": "reference",
                "typeName": "Message4",
              },
              "type": "arrayElement",
            },
            "type": "array",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
    "Message4": Object {
      "members": Array [
        Object {
          "name": "x",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "reference",
        "typeName": "Message3",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel interface.ts getInterface4: getInterface4 1`] = `
Object {
  "deps": Object {
    "Message5": Object {
      "members": Array [
        Object {
          "name": "y",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "x",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message5",
  },
}
`;

exports[`generateParserModel interface.ts getInterface5: getInterface5 1`] = `
Object {
  "deps": Object {
    "Message6": Object {
      "members": Array [
        Object {
          "name": "x",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "y",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message6",
  },
}
`;

exports[`generateParserModel literal.ts getBooleanLiteral1: getBooleanLiteral1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "literal": true,
    "type": "boolean-literal",
  },
}
`;

exports[`generateParserModel literal.ts getBooleanLiteral2: getBooleanLiteral2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "literal": true,
        "type": "boolean-literal",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel literal.ts getNumberLiteral1: getNumberLiteral1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "literal": 123,
    "type": "number-literal",
  },
}
`;

exports[`generateParserModel literal.ts getNumberLiteral2: getNumberLiteral2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "literal": 123,
        "type": "number-literal",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel literal.ts getStringLiteral1: getStringLiteral1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "literal": "aap",
    "type": "string-literal",
  },
}
`;

exports[`generateParserModel literal.ts getStringLiteral2: getStringLiteral2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "element": Object {
      "parser": Object {
        "literal": "aap",
        "type": "string-literal",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel native.ts getBoolean: getBoolean 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "type": "boolean",
  },
}
`;

exports[`generateParserModel native.ts getNumber: getNumber 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "type": "number",
  },
}
`;

exports[`generateParserModel native.ts getString: getString 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "type": "string",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias1: getTypeAlias1 1`] = `
Object {
  "deps": Object {
    "Message1": Object {
      "members": Array [
        Object {
          "name": "s",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "n",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "b",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "os",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "on",
          "optional": true,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "ob",
          "optional": true,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "nested",
          "optional": false,
          "parser": Object {
            "type": "reference",
            "typeName": "NestedMessage1",
          },
          "type": "member",
        },
        Object {
          "name": "oNested",
          "optional": true,
          "parser": Object {
            "type": "reference",
            "typeName": "NestedMessage1",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
    "NestedMessage1": Object {
      "members": Array [
        Object {
          "name": "s",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "n",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "b",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
        Object {
          "name": "os",
          "optional": true,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "on",
          "optional": true,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "ob",
          "optional": true,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message1",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias2: getTypeAlias2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "type": "string",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias3: _getTypeAlias3 1`] = `
Object {
  "deps": Object {
    "Message3": Object {
      "element": Object {
        "parser": Object {
          "type": "number",
        },
        "type": "arrayElement",
      },
      "type": "array",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message3",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias3: getTypeAlias3 1`] = `
Object {
  "deps": Object {
    "Message3": Object {
      "element": Object {
        "parser": Object {
          "type": "number",
        },
        "type": "arrayElement",
      },
      "type": "array",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message3",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias4: getTypeAlias4 1`] = `
Object {
  "deps": Object {
    "Message3": Object {
      "element": Object {
        "parser": Object {
          "type": "number",
        },
        "type": "arrayElement",
      },
      "type": "array",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "Message3",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias5: getTypeAlias5 1`] = `
Object {
  "deps": Object {
    "Message3": Object {
      "element": Object {
        "parser": Object {
          "type": "number",
        },
        "type": "arrayElement",
      },
      "type": "array",
    },
  },
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "reference",
        "typeName": "Message3",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel typeAlias.ts getTypeAlias6: getTypeAlias6 1`] = `
Object {
  "deps": Object {
    "Message3": Object {
      "element": Object {
        "parser": Object {
          "type": "number",
        },
        "type": "arrayElement",
      },
      "type": "array",
    },
    "Message5": Object {
      "members": Array [
        Object {
          "name": "x",
          "optional": false,
          "parser": Object {
            "element": Object {
              "parser": Object {
                "type": "reference",
                "typeName": "Message3",
              },
              "type": "arrayElement",
            },
            "type": "array",
          },
          "type": "member",
        },
        Object {
          "name": "y",
          "optional": false,
          "parser": Object {
            "element": Object {
              "parser": Object {
                "type": "string",
              },
              "type": "arrayElement",
            },
            "type": "array",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "element": Object {
      "parser": Object {
        "type": "reference",
        "typeName": "Message5",
      },
      "type": "arrayElement",
    },
    "type": "array",
  },
}
`;

exports[`generateParserModel union.ts getUnion1: getUnion1 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "oneOf": Array [
      Object {
        "type": "string",
      },
      Object {
        "type": "number",
      },
      Object {
        "literal": false,
        "type": "boolean-literal",
      },
      Object {
        "literal": true,
        "type": "boolean-literal",
      },
    ],
    "type": "union",
  },
}
`;

exports[`generateParserModel union.ts getUnion2: getUnion2 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "oneOf": Array [
      Object {
        "type": "string",
      },
      Object {
        "type": "number",
      },
      Object {
        "literal": false,
        "type": "boolean-literal",
      },
      Object {
        "literal": true,
        "type": "boolean-literal",
      },
    ],
    "type": "union",
  },
}
`;

exports[`generateParserModel union.ts getUnion3: getUnion3 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "oneOf": Array [
      Object {
        "literal": false,
        "type": "boolean-literal",
      },
      Object {
        "literal": "string",
        "type": "string-literal",
      },
      Object {
        "literal": 123,
        "type": "number-literal",
      },
    ],
    "type": "union",
  },
}
`;

exports[`generateParserModel union.ts getUnion4: getUnion4 1`] = `
Object {
  "deps": Object {},
  "root": Object {
    "oneOf": Array [
      Object {
        "element": Object {
          "parser": Object {
            "type": "string",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      Object {
        "element": Object {
          "parser": Object {
            "type": "number",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
      Object {
        "element": Object {
          "parser": Object {
            "type": "boolean",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
    ],
    "type": "union",
  },
}
`;

exports[`generateParserModel union.ts getUnion5: getUnion5 1`] = `
Object {
  "deps": Object {
    "X1": Object {
      "members": Array [
        Object {
          "name": "a",
          "optional": false,
          "parser": Object {
            "type": "string",
          },
          "type": "member",
        },
        Object {
          "name": "b",
          "optional": false,
          "parser": Object {
            "type": "number",
          },
          "type": "member",
        },
        Object {
          "name": "c",
          "optional": false,
          "parser": Object {
            "type": "boolean",
          },
          "type": "member",
        },
      ],
      "type": "object",
    },
  },
  "root": Object {
    "oneOf": Array [
      Object {
        "type": "reference",
        "typeName": "X1",
      },
      Object {
        "element": Object {
          "parser": Object {
            "type": "reference",
            "typeName": "X1",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
    ],
    "type": "union",
  },
}
`;

exports[`generateParserModel union.ts getUnion6: getUnion6 1`] = `
Object {
  "deps": Object {
    "X2": Object {
      "oneOf": Array [
        Object {
          "literal": false,
          "type": "boolean-literal",
        },
        Object {
          "literal": 123,
          "type": "number-literal",
        },
        Object {
          "literal": "str",
          "type": "string-literal",
        },
      ],
      "type": "union",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "X2",
  },
}
`;

exports[`generateParserModel union.ts getUnion7: getUnion7 1`] = `
Object {
  "deps": Object {
    "X3": Object {
      "oneOf": Array [
        Object {
          "members": Array [
            Object {
              "name": "a",
              "optional": false,
              "parser": Object {
                "type": "string",
              },
              "type": "member",
            },
          ],
          "type": "object",
        },
        Object {
          "members": Array [
            Object {
              "name": "b",
              "optional": false,
              "parser": Object {
                "type": "number",
              },
              "type": "member",
            },
          ],
          "type": "object",
        },
        Object {
          "members": Array [
            Object {
              "name": "c",
              "optional": false,
              "parser": Object {
                "type": "boolean",
              },
              "type": "member",
            },
          ],
          "type": "object",
        },
      ],
      "type": "union",
    },
  },
  "root": Object {
    "type": "reference",
    "typeName": "X3",
  },
}
`;

exports[`generateParserModel union.ts getUnion8: getUnion8 1`] = `
Object {
  "deps": Object {
    "X3": Object {
      "oneOf": Array [
        Object {
          "members": Array [
            Object {
              "name": "a",
              "optional": false,
              "parser": Object {
                "type": "string",
              },
              "type": "member",
            },
          ],
          "type": "object",
        },
        Object {
          "members": Array [
            Object {
              "name": "b",
              "optional": false,
              "parser": Object {
                "type": "number",
              },
              "type": "member",
            },
          ],
          "type": "object",
        },
        Object {
          "members": Array [
            Object {
              "name": "c",
              "optional": false,
              "parser": Object {
                "type": "boolean",
              },
              "type": "member",
            },
          ],
          "type": "object",
        },
      ],
      "type": "union",
    },
  },
  "root": Object {
    "oneOf": Array [
      Object {
        "members": Array [
          Object {
            "name": "a",
            "optional": false,
            "parser": Object {
              "type": "string",
            },
            "type": "member",
          },
        ],
        "type": "object",
      },
      Object {
        "members": Array [
          Object {
            "name": "b",
            "optional": false,
            "parser": Object {
              "type": "number",
            },
            "type": "member",
          },
        ],
        "type": "object",
      },
      Object {
        "members": Array [
          Object {
            "name": "c",
            "optional": false,
            "parser": Object {
              "type": "boolean",
            },
            "type": "member",
          },
        ],
        "type": "object",
      },
      Object {
        "element": Object {
          "parser": Object {
            "type": "reference",
            "typeName": "X3",
          },
          "type": "arrayElement",
        },
        "type": "array",
      },
    ],
    "type": "union",
  },
}
`;
